### PR TITLE
Add collector for TrackJS

### PR DESCRIPTION
**Pull request by sdawson**
This collector uses the TrackJS API to collect metrics about the number of errors triggered in r101-frontend in the previous minute (or whatever time interval `resolution` is set to).

The TrackJS API (http://docs.trackjs.com/data-api/) provides a limited set of different endpoints, and since the smallest time interval you can get data over is by day, we instead use the default `/v2/errors` with a 250 result page size to count the number of new errors seen in the last minute.

The 250 result page size was recommended by one of the TrackJS engineers when I enquired about rate limits for the service. By default the `/v2/errors` endpoint returns all the errors ever, so to avoid trying to count all of these up (and to limit the number of errors counted every time this collector is started) there's a max page limit set. Every minute, the collector will make a request to `/v2/errors`, and count the number of errors returned until either a error previously counted is seen, the collector hits the max page limit, or the collector reads all of the errors returned by the endpoint, which would only happen if < 250 errors are returned in total.

The page size and max number of pages are adjustable, and can be changed once the typical number of errors per minute is known.

Test Plan: This collector was tested by running r101-metrics-collector locally, logging all the errors returned, and ensuring that each of the cases that stop the collector from counting further worked as expected.

### CODE REVIEW

**Comment by turadg** [comment]
❤️ 

Can we have a unit test on some sample responses from TrackJS? Would make it much easier for anyone to contribute to this code in the future.

**Comment by sdawson** [comment]
Good idea. I should add a WIP to this since I'm still testing edge cases, but I'll bundle those into unit tests so we can repro them easily when extending or modifying this.

**Comment by dalyons** [comment]
looks good, well tested, neat use of `Enumerator`. 

Stylistic nitpics;
- ruby never uses CamelCase
- prefer named args w/defaults (`limit: 500`) over default positional args ( `limit = 500` )

File: lib/collective/collectors/trackjs.rb [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    private
+
+    def instrument_errors
+      count = 0
+      application = ""
+      paged("/#{customer_id}/v1/errors").each do |error|
+        application = error['application'] if count == 0
+        count += 1
+      end
+      instrument 'trackjs.url.errors', count, source: application, type: 'count'
```
Comment (sanjayprabhu) [inline lib/collective/collectors/trackjs.rb]: This url returns errors for all applications right? Seems like you might want to either filter by application in the url, or count the errors by application and call instrument multiple times.

Comment (turadg) [inline lib/collective/collectors/trackjs.rb]: Thanks for catching that. We want to track r101-marketing separately from r101-frontend errors.

File: lib/collective/collectors/trackjs.rb [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    # 1. We see an error we've already returned
+    # 2. We hit a maximum page limit (to avoid scanning the entire list of
+    #    errors the first time the collector is run)
+    # 3. We hit the very end of the error list
+    #
+    def paged(path, params={})
+      Enumerator.new do |yielder|
+        page = 1
+        pageSize = 250
+        maxPages = 1
```
Comment (sanjayprabhu) [inline lib/collective/collectors/trackjs.rb]: What if there's more than one page of errors for the past 60 seconds? Also, if someone changes the frequency, they might not think to change thing. Maybe use the `timestamp` and look back upto the last seen timestamp? You could default that to `currenttime - frequency`

File: lib/collective/collectors/trackjs.rb (hunk 1 above)
Comment (sanjayprabhu) [inline lib/collective/collectors/trackjs.rb]: Ok, yeah then I imagine it might also be useful to break the errors down by application and browser name/version. Take a look at https://github.com/remind101/collective/blob/master/lib/collective/collectors/postgres.rb#L22-L23.

This allows some powerful visualizations like the last last four graphs here: https://app.datadoghq.com/dash/87613/whats-happening-w-prod-api-db?live=true&page=0&is_auto=false&from_ts=1461884463516&to_ts=1462489263516&tile_size=m

File: spec/support/trackjs_helpers.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    return {
+      "message" => params[:message],
+      "timestamp" => params[:timestamp],
+      "url" => params[:url],
+      "id" => params[:id],
+      "browserName" => "Chrome",
+      "browserVersion" => "49.0.2623",
+      "entry" => "window",
+      "application" => params[:application],
```
Comment (sanjayprabhu) [inline spec/support/trackjs_helpers.rb]: I didn't see this in the docs: http://docs.trackjs.com/data-api/errors. Is it actually being returned or are the docs incorrect?

File: lib/collective/collectors/trackjs.rb [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    end
+
+    private
+
+    def instrument_errors
+      count = 0
+      application = ""
+      paged("/#{customer_id}/v1/errors").each do |error|
+        application = error['application'] if count == 0
+        count += 1
```
Comment (sanjayprabhu) [inline lib/collective/collectors/trackjs.rb]: Also, do you know if the "Time seen" header is a property of the error or are they summing up the errors by the Message? Seems like the latter, but would be good to confirm.

File: spec/support/trackjs_helpers.rb (hunk 3 above)
Comment (sdawson) [inline spec/support/trackjs_helpers.rb]: These stubs are actually just edited versions of results I got from the real API, so their docs are incorrect in this case (the real API does return the application in each error).

File: lib/collective/collectors/trackjs.rb (hunk 1 above)
Comment (sdawson) [inline lib/collective/collectors/trackjs.rb]: Weirdly when I originally wrote this I looked at our trackjs application settings in their web app and only saw an entry for r101-frontend, so I didn't do any application filtering in this version. I can add it though, since it'd be a good idea to have that sort of filtering ability build in by default anyway.

File: lib/collective/collectors/trackjs.rb (hunk 2 above)
Comment (sdawson) [inline lib/collective/collectors/trackjs.rb]: This was intentional, since it's assumed that we'd start running this for a while, get an understanding of what a normal error rate for a particular collection frequency looks like then adjust accordingly, but using `currenttime - frequency` to prevent running over all the errors when the collector is first turned on sounds like a good way to do this without having to tune it so much initially :)

File: lib/collective/collectors/trackjs.rb [hunk 5]
```diff
@@ -5,25 +5,25 @@ class TrackJS < Collective::Collector
...
+    def initialize(options = {})
+      @last_seen_id = options[:last_seen_id] || {}
       super(options)
     end
 
     collect do
-      instrument_errors
+      instrument_errors "r101-frontend"
+      instrument_errors "r101-marketing"
     end
```
Comment (sanjayprabhu) [inline lib/collective/collectors/trackjs.rb]: This should be configurable via the `Collectfile`, don't hardcode here.

**Comment by sanjayprabhu** [comment]
👍

**Comment by turadg** [comment]
:tada:
